util/EIP-152/benches/bench.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+
+use criterion::{Criterion, criterion_group, criterion_main, black_box, Throughput, BenchmarkId};
+use std::mem;
+use std::sync::atomic::{AtomicPtr, Ordering};
+use eip_152::portable;
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+use eip_152::avx2;
+
+type FnRaw = *mut ();
+type Blake2bF = fn(&mut [u64; 8], [u64; 16], [u64; 2], bool, usize);
+
+static FN: AtomicPtr<()> = AtomicPtr::new(detect as FnRaw);
+
+fn detect(state: &mut [u64; 8], message: [u64; 16], count: [u64; 2], f: bool, rounds: usize) {
+	let fun = if is_x86_feature_detected!("avx2") {
+		avx2::compress as FnRaw
+	} else {
+		portable::compress as FnRaw
+	};
+	FN.store(fun as FnRaw, Ordering::Relaxed);
+	unsafe {
+		mem::transmute::<FnRaw, Blake2bF>(fun)(state, message, count, f, rounds)
+	}
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn avx_ifunc_benchmark(c: &mut Criterion) {
+	let mut group = c.benchmark_group("avx_ifunc_impl");
+
+	for rounds in 12..15 {
+		group.throughput(Throughput::Elements(rounds as u64));
+		group.bench_with_input(
+			BenchmarkId::from_parameter(rounds),
+			&rounds,
+			|b, rounds| {
+				b.iter(|| {
+					let mut state = [
+						0x6a09e667f2bdc948_u64, 0xbb67ae8584caa73b_u64,
+						0x3c6ef372fe94f82b_u64, 0xa54ff53a5f1d36f1_u64,
+						0x510e527fade682d1_u64, 0x9b05688c2b3e6c1f_u64,
+						0x1f83d9abfb41bd6b_u64, 0x5be0cd19137e2179_u64,
+					];
+
+					let message = [
+						0x0000000000636261_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+						0x0000000000000000_u64, 0x0000000000000000_u64,
+					];
+					let count = [3, 0];
+					let f = true;
+
+					unsafe {
+						let fun = FN.load(Ordering::Relaxed);
+						mem::transmute::<FnRaw, Blake2bF>
+							(fun)
+							(
+								black_box(&mut state),
```

Comment:
`blackbox` reduces noise from comiler optimizartions

Reply:
Not sure what that means, can you educate me? How would I measure the impact on noise (other than seeing a change in the benchmark numbers)?

Reply:
> A function that is opaque to the optimizer, used to prevent the compiler from optimizing away computations in a benchmark.

From what I can tell, `black_box` prevents compiler optimizations from interfering with the benchmark, removing it gives me this message: 

![Screenshot from 2019-10-01 15-21-42](https://user-images.githubusercontent.com/16494314/65970876-35f84580-e45f-11e9-81be-3c5e78ae1a6c.png)

Reply:
I get that message both with and without using `black_box`. I think that message is simply Criterion-speak for "noticed some changes compared to last run, but too small to report as improvment/regression". See https://docs.rs/criterion/0.2.4/criterion/struct.Criterion.html#method.noise_threshold

Reply:
Maybe this is a better ref: https://github.com/bheisler/criterion.rs/blob/master/book/src/user_guide/command_line_output.md#change

Reply:
/cc @ordian can you guide us here? :)

Reply:
`black_box` is a hint to the compiler to prevent it from optimizing the code away. The message you're seeing about `Change within noise threshold` is likely to be caused by other factors, like other processes running on your cpu, cpu throttling after compiling some rust code, etc. Unless you're seeing this consistently only with `black_box` removed, which I doubt.

IMHO, it's fine to keep `black_box` even if the compiler doesn't optimize the code away w/o it as of now.

Reply:
> IMHO, it's fine to keep `black_box` even if the compiler doesn't optimize the code away w/o it as of now.

Two questions on this:

1. usually I see `black_box` used around the *function* under test, not its params. I think both usages are fine, but I think Criterions `black_box` is not zero-overhead so perhaps it introduces less noise to apply it only once rather than one time for each param? i.e. use it like this `black_box(myfun(1, "abc"))` rather than `myfun(black_box(1), black_box("abc"))`?

2. removing the blackboxes I see a sizeable perf increase here, 10 - 25%. It seems a bit too little to be explained with a const optimization and a bit much for `black_box` overhead. I did not run with a silent machine though so not sure how to interpret this. :/ Not really a question I guess.

Reply:
1. I don't know what's the best practice here, but the getting started page of criterion https://bheisler.github.io/criterion.rs/book/getting_started.html uses black_box on params. 
2. Sometimes I see measurable perf, but then I realize that rls was running previously. So making sure it's a silent run and the results are reproducible is important.